src/Main.elm

```diff
@@ -1,10 +1,829 @@
+
+-- 3. is a magma since the operation between these two things returns another thing
+-- of the same type
+
+-- we use this to "nest" circles into one another
+nest : E.Element Msg -> E.Element Msg -> E.Element Msg
+nest thing1 thing2 =
+    -- we can't use "E.row" or "E.column" since we don't want them displayed in
+    -- a row or a column, but one in front of the other
+    E.el
+        [ E.inFront thing2
+        , E.centerY
+        , E.centerX
+        ]
+        <| thing1
+
+
+
+-- in a production setting I would actually use svg for this sort of thing
+
+
+viewCircle : E.Color -> Float -> Float -> Float -> E.Element Msg
+viewCircle color_ width startingPoint currentPosition =
+    let
+        w =
+            round <| width * abs (sin (currentPosition * startingPoint))
+    in
+    E.el
+        [ E.width <| E.px (round width)
+        , E.height <| E.px (round width)
+        ]
+        <| E.el
+            [ EBackground.color color_
+            , E.width <| E.px w
+            , E.height <| E.px w
+            , E.centerX
+            , E.centerY
+            , E.htmlAttribute <| Html.Attributes.style "border-radius" "50%"
+            ]
+            <| E.none
+
+
+
+{-
+   Note on error reporting:
+   When I saw these ids that represent "parents" or "folders" I thought a lot
+   could go wrong.
+   What if we have a folder whose parent id actually points to an "audience"? Or
+   an audience whose "folder" id actually points to another "audience"?
+   What if there there's cyclical references? folder 1 points to folder 2 and vice versa?
+   What if we have a "parent" or "folder" id that points to a folder or parent that
+   doesn't exist?
+
+   Initially I thought "ok, cool maybe we can do something in elm to address this".
+   But then I realized that that would probably make the app much slower. Having to
+   to check for all of these potential errors would bring a lot of complexity
+   code-wise and also time-complexity wise. So instead I concluded that this may be
+   better addressed on the backend by writing very good tests.
+   This way our frontend is faster, easier to understand and our tests for these issues
+   only live in one place: the backend.
+
+   Additionally, this made me think that since there are so many things that could go
+   wrong by having this model of references between the audiences and audienceFolders,
+   maybe this would even better be addressed not by writing really good tests on the
+   backend, but instead by data modelling and representing these relationships in a
+   better way, thus making these invalid states impossible to even represent, so we
+   wouldn't even need to test for them.
+
+   Relevant: Richard Feldman's talk "Making Impossible States Impossible"
+   https://www.youtube.com/watch?v=IcgmSRJHu_8
+
+   Just for the record: In production code I wouldn't write these sorts of very long
+   comments. I'd instead probably open up an issue or ping the team on slack.
```

Comment:
👍 for your thoughts here. Anyway what about empty folders? Do you think about them? In your implementation are empty folders always shown even if they are clearly belongs to `Currated` audiences only.
Also about complexity and covering impossible states, did you consider use some precomputed data structure, where you will be constructing whole `tree` only once (when data are loaded) instead of checking if audience belongs to currently selected folders in every view cycle?

Reply:
I have some code I want to show regarding this, but I wanna make sure it's right, so I'll answer this tomorrow because it's late right now